designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala

```diff
@@ -0,0 +1,160 @@
+package pl.touk.nussknacker.ui.validation
+
+import cats.data.Validated.Invalid
+import io.circe.{Decoder, Encoder, HCursor, Json}
+import pl.touk.nussknacker.engine.api.NodeId
+import pl.touk.nussknacker.engine.api.context.PartSubGraphCompilationError
+import pl.touk.nussknacker.engine.api.deployment.CustomActionCommand
+import pl.touk.nussknacker.engine.api.parameter.ParameterName
+import pl.touk.nussknacker.engine.deployment.{
+  CustomActionDefinition,
+  CustomActionParameter,
+  CustomActionValidationResult
+}
+import pl.touk.nussknacker.engine.graph.expression.Expression
+import pl.touk.nussknacker.restmodel.CustomActionRequest
+import pl.touk.nussknacker.ui.BadRequestError
+
+class CustomActionValidator(val allowedActions: List[CustomActionDefinition]) {
+
+  def validateCustomActionParams(
+      request: CustomActionRequest
+  ): Either[CustomActionValidationError, CustomActionValidationResult] = {
```

Comment:
Too much happens here. It seems that with `Either[CustomActionValidationError, CustomActionValidationResult]` you have 2 levels of error handling:
- Left(CustomActionValidationError) is returned when validation errors are detected
- Right(CustomActionValidationResult) contains validation errors.

Here we should return something simple like:
```
CustomActionValidationResult(parameterErrors: List[NodeValidationError])
```
In the code below we should handle validation error as Validated/ValidatedNel, instead of exception (see CustomActionValidationError and MismatchedParamsError). Also, keep in mind that:
- no validation error = ok, goes straight to 200 OK
- some validation errors = still ok, we should return list of invalid fields with reason of invalidity
- technical exception = not ok, goes straight to 500 server error (?)

Reply:
Extracted error logic to ManagementApiHttpService, CustomActionValidator now returns ValidatedNel[SubGraphCompilationError, Unit]